<?xml version="1.0"?>
<robot
    xmlns:xacro="http://www.ros.org/wiki/xacro" name="hydrus" >

  <!-- basic kinematics model -->
  <xacro:include filename="$(find hydrus)/urdf/common.xacro" />
  <xacro:include filename="$(find hydrus)/urdf/link.urdf.xacro" />

  <xacro:hydrus_link links="4" self="1" rotor_direction="-1" with_battery = "0" />
  <xacro:hydrus_link links="4" self="2" rotor_direction="1"  with_battery = "0" />
  <xacro:hydrus_link links="4" self="3" rotor_direction="-1" with_battery = "0" />
  <xacro:hydrus_link links="4" self="4" rotor_direction="1"  with_battery = "0" />

  <!-- special battery arrangement -->
  <xacro:extra_module name = "bat1" parent = "link1" visible = "1"
                      model_url = "package://hydrus/urdf/mesh/battery/Kypom-3000-6s.dae" >
    <origin xyz="${link_length/2} 0.0 -0.048" rpy="0 0 0"/>
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <mass value="0.4108"/>
      <inertia
          ixx="0.0001" iyy="0.0006" izz="0.0006"
          ixy="0.0" ixz="0.0"  iyz="0.0"/>
    </inertial>
  </xacro:extra_module>

  <xacro:extra_module name = "bat2" parent = "link4" visible = "1"
                      model_url = "package://hydrus/urdf/mesh/battery/Kypom-3000-6s.dae" >
    <origin xyz="${link_length/2} 0.0 -0.048" rpy="0 0 0"/>
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <mass value="0.4108"/>
      <inertia
          ixx="0.0001" iyy="0.0006" izz="0.0006"
          ixy="0.0" ixz="0.0"  iyz="0.0"/>
    </inertial>
  </xacro:extra_module>

  <!-- onboard -->
  <!-- 1.  processor -->
  <!-- 1.1 flight controller -->
  <xacro:extra_module name = "fc" parent = "link2" visible = "1"
                      model_url = "package://hydrus/urdf/mesh/ver2/flight_controller/spinal.dae"> <!-- same with intel euclid -->
    <origin xyz="${link_length / 2 + 0.2221} ${-4.4*0.001} 0.02135" rpy="0 0 0"/>
    <inertial>
      <mass value = "0.05" />
      <origin xyz="${13*0.001} ${4.4*0.001} ${-9*0.001}" rpy="0 0 0"/>
      <inertia
          ixx="0.00001" ixy="0.0" ixz="0.0"
          iyy="0.00001" iyz="0.0"
          izz="0.00002"/>
    </inertial>
  </xacro:extra_module>
  <!-- end: flight controller -->

  <!-- 1.2 processor: jetson tx2 -->
  <!-- confirm the inertial parameter with real machine -->
  <xacro:extra_module name = "pc" parent = "link3" visible = "1"
                      model_url = "package://hydrus/urdf/mesh/ver2/processor/jetson_tx2_unit.dae" >
    <origin xyz="${link_length / 2 - 0.23775} 0.0 ${-0.01}" rpy="0 0 0"/>
    <inertial>
      <mass value = "0.181" />
      <origin xyz="${18 * 0.001} ${4 * 0.001} ${-18 * 0.001}" rpy="0 0 0"/>
      <inertia
          ixx="0.00007" ixy="0.0" ixz="0.0"
          iyy="0.00003" iyz="0.0"
          izz="0.00009"/>
    </inertial>
  </xacro:extra_module>
  <!-- end: processor: jetson tx2 -->

  <!-- 2.  sensor -->
  <!-- 2.1 leddar one -->
  <xacro:extra_module name = "leddarone" parent = "link3" visible = "0"
                      model_url = "package://hydrus/urdf/mesh/sensor/leddar_one_tx2_attached_mode.dae">
    <origin xyz="${link_length/2 - 0.1855} 0.0 -0.0626" rpy="${pi} 0 0"/>
    <inertial>
      <origin xyz="-0.008 0.0 0.0" rpy="0 0 0"/>
      <mass value="0.025"/>
      <inertia
          ixx="0.00001" iyy="0.000006" izz="0.00001"
          ixy="0.000000" ixz="0.000000" iyz="0.000000"/>
    </inertial>
  </xacro:extra_module>
  <!-- end: leddar one -->

  <!-- 2.2 gps -->
  <xacro:extra_module name = "gps" parent = "link2" visible = "0"
                      model_url = "package://hydrus/urdf/mesh/sensor/gps_ublox_m8n.dae">
    <origin xyz="${link_length/2 + 0.1925} 0.0 0.152" rpy="0 0 0"/>
    <inertial>
      <origin xyz="0.000000 0.000000 -0.013" rpy="0 0 0"/>
      <mass value="0.042"/>
      <inertia
          ixx="0.00006" iyy="0.00006" izz="0.000007"
          ixy="0.000000" ixz="0.000000" iyz="0.000000"/>
    </inertial>
  </xacro:extra_module>
  <xacro:extra_module name = "magnet" parent = "gps">
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <inertial>
      <mass value = "0.00001" />
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia
          ixx="0.000001" ixy="0.0" ixz="0.0"
          iyy="0.000001" iyz="0.0"
          izz="0.000002"/>
    </inertial>
  </xacro:extra_module>
  <!-- end: gps -->

  <xacro:property name="body_radius" value="0.50" />
  <material name="Dummy_body_color">
    <color rgba="1.0 1.0 1.0 0.5"/>
  </material>

  <link name="dummy_body">
    <inertial>
      <mass value = "0.0001" />
      <origin xyz="0 0 0 " rpy="0 0 0"/>
      <inertia
          ixx="0.000001" ixy="0.0000" ixz="0.0000"
          iyy="0.000001" iyz="0.00000"
          izz="0.000001"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.30" radius="${body_radius}"/>
      </geometry>
      <material name="Dummy_body_color"/>
    </visual>
    <!-- <collision>
        <origin rpy="0 0 0" xyz="0 0 0"/>
        <geometry>
          <cylinder length="0.02" radius="${body_radius}"/>
        </geometry>
    </collision> -->
  </link>

  <joint name="dummy_joint_1" type="fixed">
    <parent link="link1"/>
    <child link="dummy_body"/>
    <origin rpy="0 0 0" xyz="${link_length* 0.5} ${link_length* 0.5} 0.0"/>
  </joint>

   <!-- <link name="dummy_body_visualize">
    <inertial>
      <mass value = "0.0001" />
      <origin xyz="0 0 0 " rpy="0 0 0"/>
      <inertia
          ixx="0.000001" ixy="0.0000" ixz="0.0000"
          iyy="0.000001" iyz="0.00000"
          izz="0.000001"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
       <mesh
        scale="2 2 2"
        filename="package://agile_multirotor/urdf/mesh/quadrotor.stl" />
      </geometry>
    </visual>
  </link>

  <joint name="dummy_joint_2" type="fixed">
    <parent link="link1"/>
    <child link="dummy_body_visualize"/>
    <origin rpy="0 0 0" xyz="${link_length* 0.5} ${link_length* 0.5} 0.0"/>
  </joint> -->

</robot>
